txnkv/transaction/2pc.go

```diff
@@ -622,8 +625,17 @@ func (c *twoPhaseCommitter) initKeysAndMutations(ctx context.Context) error {
 				// Do not exit immediately here. To rollback the pessimistic locks (if any), we need to finish
 				// collecting all the keys.
 				// Keep only the first assertion error.
+
+				// assertion errors is treated differently from other errors. If there is an assertion error,
+				// it's probably cause by loss of pessimistic locks, so we can't directly return the assertion error.
+				// Instead, we stash the error, forbid async commit and 1PC, then let the prewrite continue.
```

Comment:
Let me confirm. Forbidding async commit and 1PC is to give us the chance to abort the transaction by the client. And the main goal of this PR is to change the assertion error in this case to become a constraint check failure or write failure. Do I understand it correctly?

Reply:
Yes exactly